### PR TITLE
Use JWTAuth Middleware to check correct authentication with the backend.

### PR DESCRIPTION
**Pull request by minuscorp**
captura de pantalla 2018-05-15 a las 13 07 55

### CODE REVIEW

**Comment by minuscorp** [comment]
I'd like to add some tests to this before merging😁

**Comment by snatchev** [comment]
@minuscorp thanks for this! Looks pretty good at first pass.

Do you have any thoughts on having this as a module instead of a class? Perhaps something that can be included in https://github.com/fastlane/ci/blob/master/app/features-json/json_controller.rb ?

I have a slight preference to composition over inheritance for this kind of behavior, but I want to be sensitive to not doing surprising things in the code. Would be happy to discuss.

**Comment by minuscorp** [comment]
My main decision on using inheritance over a module was to follow the same design pattern that was used in the erb application, so contributors would see the same behavior and code throughout the codebase. Also, using inheritance, we kind of lock people to use the resources given by the parent class and prevent future controllers to inherit from foreign or extraneous classes that might be out of our control

**Comment by snatchev** [comment]
@minuscorp okay, thank you for the explanation. Sounds good to me.

**Comment by minuscorp** [comment]
@snatchev help with the tests! 🆘

File: spec/features-json/jwt_auth_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      expect(response.status).to eql(401)
+    end
+  end
+
+  context "Client makes a request with an expired authentication token" do
+    let(:request) { Rack::MockRequest.new(middleware) }
+    let(:expired_header) { { "HTTP_AUTHORIZATION": "Bearer eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjExNjM5MjYxLCJpYXQiOjExNjM5MjYxLCJpc3MiOiJmYXN0bGFuZS5jaSJ9._wzA6VzEuex1wJZctYHk94lCDMydOIe7scENvsCqTes" } }
+    let(:response) { request.get("/buy/tacos", expired_header) }
+
+    it "Returns a 403 status" do
```
[the remark this answers is not in the document]
Comment (minuscorp) [reply spec/features-json/jwt_auth_spec.rb]: Totally, should change it on JwtAuth tho, as now it is 403

[the remark this answers is not in the document]
Comment (nakhbari) [reply spec/features-json/jwt_auth_spec.rb]: Ya that would be good to get that in now, so we don't have to worry about it later 👍

File: app/features-json/json_authenticated_controller_base.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  # A user is required to be logged in to access any routes on this controller
+  #
+  class JSONAuthenticatedControllerBase < ControllerBase
+    use FastlaneCI::JwtAuth
+    include FastlaneCI::Logging
+
+    # Method to obtain the current user given their primary key.
+    # Raises if no User found.
+    # @return [User]
+    def user!
```
Comment (snatchev) [inline app/features-json/json_authenticated_controller_base.rb]: @minuscorp for these kinds of find-or-raise patterns, I think we should start making use of Sinatra's `halt` method. As described here: http://myronmars.to/n/dev-blog/2012/01/why-sinatras-halt-is-awesome

This way, it's not raising an exception that must be caught, and this doesn't need to be a bang method.

Also, could we call it `current_user` :)

**Comment by minuscorp** [comment]
I changed the signature of the method, and halt with a 500 when user is not found. What do you think @snatchev

**Pull request merged**